File: packages/agentenkit/test/flow.test.ts
import { describe, expect, it } from 'bun:test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { bindStorage } from '../src/core/state.js';
import { contextBudget } from '../src/core/context.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryStorage } from '../src/adapters/memory.js';
import type { AgentConfig } from '../src/core/types.js';
import { resolveConfig } from '../src/core/types.js';
import type { RuntimeOptions } from '../src/ports/runtime.js';
import type { BoundStorage } from '../src/core/state.js';

const MODEL_WINDOWS: Record<string, number> = {
  'gpt-4o': 128_000,
  'claude-3-5-sonnet': 200_000,
  'gemini-1.5-pro': 265_000,
};

async function makeDeps(config: Partial<AgentConfig> = {}) {
  const storage = new MemoryStorage();
  const bus = new MemoryBus();
  const queue = new MemoryQueue();
  const kv = new MemoryKv();
  const deps: RuntimeOptions = {
    storage,
    // Isolated per test: the default store is a file on disk.
    admin: new MemoryAdminStore(),
    bus,
    queue,
    kv,
    resolveModel: (name) => ({
      instance: () => {
        throw new Error('no-llm');
      },
      contextWindow: MODEL_WINDOWS[name] ?? 128_000,
    }),
    config: resolveConfig(config),
  };
  return { deps, store: bindStorage(storage, { state: {} }), runtime: await setupAgentCore(deps), storage, bus, queue, kv };
}

/** Simulate a thread parked mid-HITL (§2.5): WAITING_FOR_INPUT + pending request. */
async function parkThread(
  r: { store: BoundStorage; kv: MemoryKv },
  toolCallId = 'call_1',
) {
  const { store, kv } = r;
  const thread = await store.threads.create(undefined);
  const seq = await kv.incr(`agent:seq:${thread.id}`);
  await store.threads.setState(thread.id, 'WAITING_FOR_INPUT');
  await kv.set(`agent:state:${thread.id}`, 'WAITING_FOR_INPUT');
  await store.events.append(thread.id, {
    threadId: thread.id,
    seq,
    type: 'INPUT_REQUIRED',
    payload: { toolCallId, toolName: 'sendEmail', arguments: { to: 'a@b.c' } },
    createdAt: new Date(),
  });
  return thread.id;
}

describe('runtime.run via handle (§5.1)', () => {
  it('creates a thread, persists the user message, marks RUNNING, enqueues', async () => {
    const { deps, store, kv, runtime, queue } = await makeDeps();
    const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });

    const res = await chat.run({ prompt: 'hello' });

    expect(res.accepted).toBe(true);
    expect(res.state).toBe('RUNNING');
    const thread = await store.threads.get(res.threadId);
    expect(thread!.state).toBe('RUNNING');
    expect(thread!.model).toBe('gpt-4o');
    const messages = await store.messages.list(res.threadId, undefined);
    expect(messages).toHaveLength(1);
    expect(messages[0]!.role).toBe('user');
    expect(queue.items).toHaveLength(1);
    expect(queue.items[0]!.threadId).toBe(res.threadId);
    expect(queue.items[0]!.agent).toBe('chat');
  });

  it('rejects a second run while one is active (§2.1 guard)', async () => {
    const { runtime } = await makeDeps();
    const chat = runtime.createStreamTextAgent({ name: 'chat' });
    const first = await chat.run({ prompt: 'a' });
    const second = await chat.run({ threadId: first.threadId, prompt: 'b' });
    expect(second.accepted).toBe(false);
    expect(second.error).toMatch(/active run/);
  });

  it('rejects when the billing pre-check fails (§4)', async () => {
    const { runtime, bus } = await makeDeps({
      billingPreCheck: async ({ state, publishEvent }) => {
        // The check can tell every client why, before the refusal lands
        await publishEvent('CREDIT_LIMIT', { org: state.orgId });
        return { ok: false, error: 'Insufficient credits' };
      },
    });
    const chat = runtime.createStreamTextAgent({ name: 'chat' });
    const res = await chat.run({ prompt: 'x', state: { orgId: 'acme' } });
    expect(res.accepted).toBe(false);
    // Both the check's own event and the platform's refusal are on the log, durably
    const limit = bus.published.find((e) => e.type === 'CREDIT_LIMIT')!;
    expect(limit.payload).toEqual({ org: 'acme' });
    const refused = bus.published.find((e) => e.type === 'RUN_REFUSED')!;
    expect(refused.payload).toEqual({ reason: 'billing', error: 'Insufficient credits' });
    expect(refused.seq).toBeGreaterThan(limit.seq);
    expect(res.error).toMatch(/Insufficient credits/);
  });

  it('threads tokenBudget through the job (§2.1)', async () => {
    const { deps, store, kv, runtime, queue } = await makeDeps();
    const chat = runtime.createStreamTextAgent({ name: 'chat' });
    await chat.run({ prompt: 'a', tokenBudget: 4_000 });
    expect(queue.items[0]!.tokenBudget).toBe(4_000);
  });

  it('lists threads most recent first (thread picker)', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const first = await store.threads.create(undefined);
    await new Promise((r) => setTimeout(r, 5)); // ensure distinct updatedAt
    const second = await store.threads.create(undefined);

    const threads = await runtime.listThreads();
    expect(threads.map((t) => t.id)).toEqual([second.id, first.id]);
  });
});

describe('runtime.stop (§5.2)', () => {
  it('writes CANCELLED to cache + durable truth and publishes STATE_CHANGE', async () => {
    const { deps, store, kv, runtime, bus } = await makeDeps();
    const chat = runtime.createStreamTextAgent({ name: 'chat' });
    const ran = await chat.run({ prompt: 'a' });

    const res = await chat.stop(ran.threadId);
    expect(res.accepted).toBe(true);
    expect(await deps.kv.get(`agent:state:${ran.threadId}`)).toBe('CANCELLED');
    const thread = await store.threads.get(ran.threadId);
    expect(thread!.state).toBe('CANCELLED');
    const last = bus.published.at(-1)!;
    expect(last.type).toBe('STATE_CHANGE');
    expect((last.payload as any).state).toBe('CANCELLED');
  });

  it('rejects stopping an IDLE thread', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const chat = runtime.createStreamTextAgent({ name: 'chat' });
    const thread = await store.threads.create(undefined);
    const res = await chat.stop(thread.id);
    expect(res.accepted).toBe(false);
    expect(res.error).toMatch(/IDLE/);
  });
});

describe('runtime.hitl.respond (§5.4)', () => {
  it('delivers an approval to the handoff key when parked (§2.5)', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const threadId = await parkThread({ store, kv });

    const res = await runtime.hitl.respond({
      threadId, toolCallId: 'call_1', approved: true, payload: { ok: 1 },
    });
    expect(res.delivered).toBe(true);
    const raw = await deps.kv.get('agent:hitl:call_1');
    expect(JSON.parse(raw!)).toEqual({ approved: true, payload: { ok: 1 } });
  });

  it('rejects an unknown toolCallId (answer-latest policy, §2.7)', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const threadId = await parkThread({ store, kv }, 'call_latest');
    const res = await runtime.hitl.respond({
      threadId, toolCallId: 'call_older', approved: true,
    });
    expect(res.delivered).toBe(false);
    expect(res.error).toMatch(/No matching/);
  });

  it('rejects a response for a thread that is not WAITING_FOR_INPUT', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const thread = await store.threads.create(undefined);
    const res = await runtime.hitl.respond({
      threadId: thread.id, toolCallId: 'call_1', approved: true,
    });
    expect(res.delivered).toBe(false);
  });
});

describe('reclaimIfOrphaned (§2.5)', () => {
  it('does nothing for a young pending request', async () => {
    const { deps, store, kv, runtime } = await makeDeps({ hitlTtlMs: 5, reclaimGraceMs: 1 });
    const thread = await store.threads.create(undefined);
    const seq = await kv.incr(`agent:seq:${thread.id}`);
    await store.threads.setState(thread.id, 'WAITING_FOR_INPUT');
    await store.events.append(thread.id, {
      threadId: thread.id, seq, type: 'INPUT_REQUIRED', payload: { toolCallId: 'c1' }, createdAt: new Date(),
    });

    expect(await runtime.hitl.reclaimIfOrphaned(thread.id)).toBe(false);
  });

  it('re-dispatches a true orphan instead of healing it inline (§2.7)', async () => {
    const { deps, store, kv, runtime, queue } = await makeDeps({ hitlTtlMs: 5, reclaimGraceMs: 1 });
    const thread = await store.threads.create(undefined);
    const seq = await kv.incr(`agent:seq:${thread.id}`);
    await store.threads.setState(thread.id, 'WAITING_FOR_INPUT');
    await kv.set(`agent:state:${thread.id}`, 'WAITING_FOR_INPUT');
    await kv.set(`agent:run:${thread.id}`, 'run-1');
    await store.events.append(thread.id, {
      threadId: thread.id, seq, type: 'INPUT_REQUIRED', payload: { toolCallId: 'c1' },
      createdAt: new Date(Date.now() - 60_000), // far older than TTL + grace
    });

    expect(await runtime.hitl.reclaimIfOrphaned(thread.id)).toBe(true);

    // It kicks the engine and touches nothing else: the engine owns the single
    // definition of what an expired approval becomes, and resolves a thread's
    // open approvals as a set rather than one at a time (§2.7).
    expect(await store.messages.list(thread.id, undefined)).toHaveLength(0);
    expect((await store.threads.get(thread.id))!.state).toBe('WAITING_FOR_INPUT');
    expect(queue.items).toHaveLength(1);
    // Resuming reuses the parked run's id (§2.1)
    expect(queue.items[0]).toMatchObject({ threadId: thread.id, runId: 'run-1' });

    // A duplicate re-dispatch is safe — the run lock and the engine's
    // readiness check make it a no-op, so it is not suppressed here.
    expect(await runtime.hitl.reclaimIfOrphaned(thread.id)).toBe(true);
    expect(queue.items).toHaveLength(2);
  });

  it('waits while ANY open approval is still answerable (§2.7)', async () => {
    const { deps, store, kv, runtime, queue } = await makeDeps({ hitlTtlMs: 5, reclaimGraceMs: 1 });
    const thread = await store.threads.create(undefined);
    await store.threads.setState(thread.id, 'WAITING_FOR_INPUT');
    for (const [id, ageMs] of [['old', 60_000], ['fresh', 0]] as const) {
      await store.events.append(thread.id, {
        threadId: thread.id,
        seq: await kv.incr(`agent:seq:${thread.id}`),
        type: 'INPUT_REQUIRED',
        payload: { toolCallId: id },
        createdAt: new Date(Date.now() - ageMs),
      });
    }

    expect(await runtime.hitl.reclaimIfOrphaned(thread.id)).toBe(false);
    expect(queue.items).toHaveLength(0);
  });
});

describe('runtime.deleteThread (§3.2)', () => {
  it('deletes the thread and everything that follows it', async () => {
    const { deps, store, kv, storage, runtime, bus } = await makeDeps();
    const thread = await store.threads.create(undefined);
    await store.messages.append(thread.id, { role: 'user', content: 'hi' });
    await store.events.append(thread.id, {
      threadId: thread.id, seq: 1, type: 'CHUNK', payload: {}, createdAt: new Date(),
    });
    await store.usage.record(thread.id, {
      agentId: null, inputTokens: 1, cachedInputTokens: 0, outputTokens: 1, totalTokens: 2,
    });
    
    await kv.set(`agent:state:${thread.id}`, 'COMPLETED');
    await kv.set(`agent:seq:${thread.id}`, '1');
    await kv.set(`agent:attempts:${thread.id}`, '2');

    const res = await runtime.deleteThread(thread.id);
    expect(res.accepted).toBe(true);

    // Cascade: messages, events, usage, runs all follow the thread
    expect(await store.threads.get(thread.id)).toBeNull();
    expect(await store.messages.list(thread.id, undefined)).toEqual([]);
    expect(await store.events.listSince(thread.id, -1)).toEqual([]);
    expect(storage.usage.recorded.filter((u) => u.threadId === thread.id)).toEqual([]);

    // Hot cache cleanup — no resurrection from kv
    for (const key of ['state', 'seq', 'attempts', 'lock']) {
      expect(await kv.get(`agent:${key}:${thread.id}`)).toBeNull();
    }

    // Live UIs are told via the bus-only notice
    expect(bus.published.at(-1)!.type).toBe('THREAD_DELETED');
  });

  it('refuses an active run but deletes a parked HITL thread (no process held, §2.5)', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const running = await store.threads.create(undefined);
    await store.threads.setState(running.id, 'RUNNING');
    const refused = await runtime.deleteThread(running.id);
    expect(refused.accepted).toBe(false);
    expect(refused.error).toMatch(/stop/i);
    expect(await store.threads.get(running.id)).not.toBeNull();

    const parked = await store.threads.create(undefined);
    await store.threads.setState(parked.id, 'WAITING_FOR_INPUT');
    expect((await runtime.deleteThread(parked.id)).accepted).toBe(true);
  });

  it('rejects an unknown thread', async () => {
    const { runtime } = await makeDeps();
    const res = await runtime.deleteThread('nope');
    expect(res.accepted).toBe(false);
    expect(res.error).toMatch(/not found/i);
  });

  it('a late resume dispatch after deletion is a no-op — no resurrection', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const thread = await store.threads.create(undefined);
    await store.threads.setState(thread.id, 'WAITING_FOR_INPUT');
    await store.events.append(thread.id, {
      threadId: thread.id, seq: 1, type: 'INPUT_REQUIRED',
      payload: { toolCallId: 'c1', toolName: 'sendEmail', resume: { agent: 'chat', model: 'gpt-4o' } },
      createdAt: new Date(),
    });

    expect((await runtime.deleteThread(thread.id)).accepted).toBe(true);
    // The §2.5 resume job arrives after the delete — nothing may come back
    await runtime.worker.handleJob({ threadId: thread.id, model: 'gpt-4o', agent: 'chat' });
    expect(await store.threads.get(thread.id)).toBeNull();
    expect(await store.messages.list(thread.id, undefined)).toEqual([]);
  });
});

describe('runtime.events (§2.2)', () => {
  it('replays since a cursor and subscribes live', async () => {
    const { deps, store, kv, runtime } = await makeDeps();
    const thread = await store.threads.create(undefined);
    for (let i = 1; i <= 3; i++) {
      await store.events.append(thread.id, {
        threadId: thread.id, seq: i, type: 'CHUNK', payload: { i }, createdAt: new Date(),
      });
    }

    const replay = await runtime.events.since(thread.id, 1);
    expect(replay.map((e) => e.seq)).toEqual([2, 3]);

    const seen: number[] = [];
    const unsubscribe = await runtime.events.subscribe(thread.id, (e) => seen.push(e.seq));
    await deps.bus.publish(thread.id, {
      threadId: thread.id, seq: 4, type: 'CHUNK', payload: {}, createdAt: new Date(),
    });
    expect(seen).toEqual([4]);
    unsubscribe();
  });
});

describe('contextBudget (§2.6)', () => {
  it('caps everything at the 265k ceiling but keeps smaller native windows', async () => {
    const { deps, store } = await makeDeps();
    // contextBudget only reads resolveModel and config, not storage.
    const ports = { ...deps, storage: store } as any;
    expect(contextBudget(ports, 'gpt-4o')).toBe(128_000);
    expect(contextBudget(ports, 'claude-3-5-sonnet')).toBe(200_000);
    expect(contextBudget(ports, 'gemini-1.5-pro')).toBe(265_000);
    expect(contextBudget(ports, 'unknown-model')).toBe(128_000); // resolveModel fallback
  });
});

File: packages/go-agentenkit/flow_test.go
package agentenkit_test

import (
	"context"
	"errors"
	"testing"
	"time"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
)

func TestRun_CreatesThreadPersistsMarksRunningEnqueues(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hello"})

	mustEqual(t, ran.State, agentenkit.StateRunning, "result state")
	if ran.RunID == "" {
		t.Fatal("no run id")
	}
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateRunning, "durable state")
	mustEqual(t, h.kvGet(agentenkit.StateKey(ran.ThreadID)), "RUNNING", "hot state")
	mustEqual(t, h.kvGet(agentenkit.RunIDKey(ran.ThreadID)), ran.RunID, "run id key")
	mustStrings(t, h.roles(ran.ThreadID), []string{"user"}, "roles")
	mustEqual(t, string(h.storage.MessageRows(ran.ThreadID)[0].Content), `"hello"`, "user content")

	jobs := h.queue.Items()
	mustEqual(t, len(jobs), 1, "jobs")
	mustEqual(t, jobs[0].Agent, "chat", "job agent")
	mustEqual(t, jobs[0].RunID, ran.RunID, "job run id")
	mustEqual(t, jobs[0].Model, "gpt-4o", "job model")
	if jobs[0].EnqueuedAt == 0 {
		t.Fatal("enqueuedAt missing")
	}
	// The user's turn goes on the bus before the run starts (§2.2)
	evs := h.events(ran.ThreadID, "")
	mustEqual(t, evs[0].Type, "MESSAGE_APPENDED", "first event")
	mustEqual(t, evs[1].Type, "STATE_CHANGE", "second event")
	rec, _ := h.admin.Runs().Get(h.ctx, ran.RunID)
	mustEqual(t, rec.State, agentenkit.StateRunning, "run record")
	mustEqual(t, rec.Prompt, "hello", "recorded prompt")
}

func TestRun_RejectsASecondRunWhileOneIsActive(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "one"})
	second, err := chat.Run(h.ctx, agentenkit.RunInput{ThreadID: ran.ThreadID, Prompt: "two"})
	if err != nil {
		t.Fatal(err)
	}
	if second.Accepted {
		t.Fatal("second run must be refused")
	}
	mustEqual(t, second.Error, "Thread has an active run", "error")
}

func TestRun_RejectsWhenTheBillingPreCheckFails(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}), func(c *agentenkit.AgentConfig) {
		c.BillingPreCheck = func(ctx context.Context, check agentenkit.BillingCheck) error {
			// The check can tell every client why, before the refusal lands
			_, _ = check.PublishEvent(ctx, "CREDIT_LIMIT", map[string]any{"org": check.State["orgId"]}, false)
			return errors.New("no credits")
		}
	})
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	res, err := chat.Run(h.ctx, agentenkit.RunInput{Prompt: "x", State: agentenkit.AgentRunState{"orgId": "acme"}})
	if err != nil {
		t.Fatal(err)
	}
	if res.Accepted {
		t.Fatal("must be refused")
	}
	mustEqual(t, res.Error, "no credits", "error")
	mustEqual(t, h.queue.Len(), 0, "nothing queued")
	// Both the app's own event and the platform's refusal are on the log
	limit := h.events(res.ThreadID, "CREDIT_LIMIT")
	mustEqual(t, len(limit), 1, "the check published")
	mustEqual(t, payload(limit[0])["org"], "acme", "with the run state")
	refused := h.events(res.ThreadID, "RUN_REFUSED")
	mustEqual(t, len(refused), 1, "RUN_REFUSED")
	mustEqual(t, payload(refused[0])["error"], "no credits", "reason")
	if refused[0].Seq == 0 {
		t.Fatal("the refusal must be durable")
	}
}

func TestRun_ThreadsTokenBudgetThroughTheJob(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "x", TokenBudget: 500})
	mustEqual(t, h.queue.Items()[0].TokenBudget, 500, "job budget")
	rec, _ := h.admin.Runs().Get(h.ctx, ran.RunID)
	mustEqual(t, *rec.TokenBudget, 500, "recorded budget")
}

func TestRun_ListsThreadsMostRecentFirst(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	a := h.run(t, chat, agentenkit.RunInput{Prompt: "a"})
	time.Sleep(2 * time.Millisecond)
	b := h.run(t, chat, agentenkit.RunInput{Prompt: "b"})
	threads, err := h.rt.ListThreads(h.ctx, nil)
	if err != nil {
		t.Fatal(err)
	}
	mustEqual(t, len(threads), 2, "threads")
	mustEqual(t, threads[0].ID, b.ThreadID, "newest first")
	mustEqual(t, threads[1].ID, a.ThreadID, "oldest last")
}

func TestStop_WritesCancelledToBothHomes(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "x"})
	res, err := chat.Stop(h.ctx, ran.ThreadID, nil)
	if err != nil || !res.Accepted {
		t.Fatalf("stop: %v %+v", err, res)
	}
	mustEqual(t, h.kvGet(agentenkit.StateKey(ran.ThreadID)), "CANCELLED", "hot")
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCancelled, "durable")
	mustStrings(t, h.states(ran.ThreadID), []string{"RUNNING", "CANCELLED"}, "states")
	// The queued job is now a no-op
	h.handleNext(t)
	mustEqual(t, h.model.Calls(), 0, "model never called")
}

func TestStop_RejectsAnIdleThread(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	th, _ := h.storage.Threads().Create(h.ctx, agentenkit.ThreadInit{}, agentenkit.StorageContext{})
	res, err := chat.Stop(h.ctx, th.ID, nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Accepted {
		t.Fatal("must be refused")
	}
	mustEqual(t, res.Error, "Cannot stop thread in state IDLE", "error")
}

func TestRespond_RejectsUnknownToolCallAndNonWaitingThread(t *testing.T) {
	h := hitlSetup(t)
	ran := h.run(t, h.chat, agentenkit.RunInput{Prompt: "delete"})
	// Not waiting yet
	res, err := h.rt.HITL.Respond(h.ctx, agentenkit.RespondInput{ThreadID: ran.ThreadID, ToolCallID: "c1", Approved: true})
	if err != nil || res.Delivered {
		t.Fatalf("expected rejection, got %+v %v", res, err)
	}
	h.handleNext(t)
	res, err = h.rt.HITL.Respond(h.ctx, agentenkit.RespondInput{ThreadID: ran.ThreadID, ToolCallID: "nope", Approved: true})
	if err != nil || res.Delivered {
		t.Fatalf("expected rejection, got %+v %v", res, err)
	}
	mustEqual(t, res.Error, "No matching pending input request", "error")
}

func TestReclaim_DoesNothingForAYoungRequest(t *testing.T) {
	h := hitlSetup(t)
	ran := h.run(t, h.chat, agentenkit.RunInput{Prompt: "delete"})
	h.handleNext(t)
	before := h.queue.Len()
	ok, err := h.rt.HITL.ReclaimIfOrphaned(h.ctx, ran.ThreadID, nil)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("must not reclaim a young request")
	}
	mustEqual(t, h.queue.Len(), before, "nothing enqueued")
}

func TestReclaim_RedispatchesATrueOrphan(t *testing.T) {
	h := hitlSetup(t, func(c *agentenkit.AgentConfig) { c.HITLTTL = 5 * time.Millisecond; c.ReclaimGrace = 0 })
	ran := h.run(t, h.chat, agentenkit.RunInput{Prompt: "delete"})
	h.handleNext(t)
	h.queue.Shift() // drop the park's own expiry to prove reclamation alone works
	time.Sleep(20 * time.Millisecond)
	ok, err := h.rt.HITL.ReclaimIfOrphaned(h.ctx, ran.ThreadID, nil)
	if err != nil || !ok {
		t.Fatalf("reclaim: %v %v", ok, err)
	}
	job := h.queue.Items()[0]
	mustEqual(t, job.RunID, ran.RunID, "reuses the run id")
	mustEqual(t, job.Agent, "chat", "rebuilt from the ticket")
	// Nothing healed inline: the thread is still parked until the worker runs
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateWaitingForInput, "state")
	h.handleNext(t)
	mustEqual(t, h.lastTerminal(ran.ThreadID)["state"], "COMPLETED", "resolved by the engine")
	mustEqual(t, len(*h.executed), 0, "expired: the tool never ran")
}

func TestDeleteThread_DeletesEverythingThatFollowsIt(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "x"})
	h.handleNext(t)
	res, err := h.rt.DeleteThread(h.ctx, ran.ThreadID, nil)
	if err != nil || !res.Accepted {
		t.Fatalf("delete: %v %+v", err, res)
	}
	th, _ := h.storage.Threads().Get(h.ctx, ran.ThreadID, agentenkit.StorageContext{})
	if th != nil {
		t.Fatal("thread still there")
	}
	mustEqual(t, len(h.storage.MessageRows(ran.ThreadID)), 0, "messages")
	evs, _ := h.storage.Events().ListSince(h.ctx, ran.ThreadID, -1, agentenkit.StorageContext{})
	mustEqual(t, len(evs), 0, "events")
	total, _ := h.storage.Usage().Total(h.ctx, ran.ThreadID, agentenkit.StorageContext{})
	mustEqual(t, total.TotalTokens, 0, "usage")
	for _, key := range []string{agentenkit.StateKey(ran.ThreadID), agentenkit.RunIDKey(ran.ThreadID), agentenkit.RunLockKey(ran.ThreadID)} {
		mustEqual(t, h.kvGet(key), "", "kv "+key)
	}
	notices := h.events(ran.ThreadID, "THREAD_DELETED")
	mustEqual(t, len(notices), 1, "THREAD_DELETED notice")
	mustEqual(t, notices[0].Seq, int64(0), "bus-only")
}

func TestDeleteThread_RefusesAnActiveRunButDeletesAParkedOne(t *testing.T) {
	h := hitlSetup(t)
	ran := h.run(t, h.chat, agentenkit.RunInput{Prompt: "delete"})
	res, _ := h.rt.DeleteThread(h.ctx, ran.ThreadID, nil)
	if res.Accepted {
		t.Fatal("must refuse a running thread")
	}
	h.handleNext(t) // parks
	res, _ = h.rt.DeleteThread(h.ctx, ran.ThreadID, nil)
	if !res.Accepted {
		t.Fatalf("parked thread must delete: %s", res.Error)
	}
	// A late resume dispatch after deletion is a no-op: no resurrection
	h.drain(t)
	th, _ := h.storage.Threads().Get(h.ctx, ran.ThreadID, agentenkit.StorageContext{})
	if th != nil {
		t.Fatal("thread resurrected")
	}
	unknown, _ := h.rt.DeleteThread(h.ctx, "nope", nil)
	mustEqual(t, unknown.Error, "Thread not found", "unknown thread")
}

func TestEvents_ReplaysSinceACursorAndSubscribesLive(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "x"})
	all, err := h.rt.Events.Since(h.ctx, ran.ThreadID, -1, nil)
	if err != nil {
		t.Fatal(err)
	}
	mustEqual(t, len(all), 2, "events so far")
	later, _ := h.rt.Events.Since(h.ctx, ran.ThreadID, all[0].Seq, nil)
	mustEqual(t, len(later), 1, "after cursor")

	var live []string
	unsub, err := h.rt.Events.Subscribe(h.ctx, ran.ThreadID, func(e agentenkit.AgentEvent) { live = append(live, e.Type) })
	if err != nil {
		t.Fatal(err)
	}
	h.handleNext(t)
	if len(live) == 0 || live[len(live)-1] != "STATE_CHANGE" {
		t.Fatalf("live tail missing, got %v", live)
	}
	_ = unsub()
}

func TestContextBudget_CapsAtTheCeilingButKeepsSmallerWindows(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	deps := h.rt.Ports(nil)
	// The resolver declares 128k for everything, which is below the ceiling
	mustEqual(t, core.ContextBudget(deps, "gpt-4o"), 128_000, "declared window")
	deps.ResolveModel = func(string) (agentenkit.ResolvedModel, error) {
		return agentenkit.ResolvedModel{}, errors.New("unknown")
	}
	mustEqual(t, core.ContextBudget(deps, "gemini-1.5-pro"), 265_000, "ceiling")
	mustEqual(t, core.ContextBudget(deps, "gpt-4o-mini"), 128_000, "table")
	mustEqual(t, core.ContextBudget(deps, "mystery"), 265_000, "default")
	deps.Config.NativeWindows = map[string]int{"mystery": 32_000}
	mustEqual(t, core.ContextBudget(deps, "mystery"), 32_000, "config table")
	_ = context.Background
}
